demo/test.c

```diff
@@ -1,5 +1,8 @@
 #include "shared.h"
 
+/* We can also test the private API here */
+#include "tommath_private.h"
```

Comment:
we can but should we?

IMO as soon as the library symbols are properly set-up you won't be able to build these tests then against the shared library

Reply:
Then we need a special test build with exposed s_* symbols. I think it is better if internal functions are tested separately. For example in the mul case we could also test the different code paths by passing the right values. But this is too fragile and I would also like to test edge cases.

This would also solve the issue with exposed CUTOFF variables. We should just make those constant and test karatsuba and toom cook separately.

Reply:
> We should just make those constant and test karatsuba and toom cook separately

IIUC they can't be made constant as they vary from platform to platform. Sure they're mostly constant now as nobody's really tweaking them but I guess that's caused by the fact that there's no proper way yet to determine them per platform. (besides running `etc/tune`)

> Then we need a special test build with exposed s_* symbols. I think it is better if internal functions are tested separately.

I'm no big fan of explicit testing of internal functionality.

> For example in the mul case we could also test the different code paths by passing the right values.

I prefer that way to make sure the entire chain works and then determine that everything is called through code coverage.

> But this is too fragile

why is this too fragile?

> I would also like to test edge cases

why would you want to test cases which can't be hit?

that all sounds a bit over the top to me

Reply:
> I'm no big fan of explicit testing of internal functionality.

I am not saying to test arbitrary internal things. However internal functions with a well defined API should be tested, e.g., mp_balance_mul. @czurnieden also added the balance mul test to the test suite, but he did not call it directly which he should have. If only the external api should be tested, there should not even be a function test_balance_mul.

> I prefer that way to make sure the entire chain works and then determine that everything is called through code coverage.

If you want that - then the test suite must be split such that internal function tests and api function tests can be compiled separately. No big deal. `#if TEST_INTERNAL .... #endif`.
Ensure that stuff is hit via code coverage is good practice, but not enough and it depends on external tooling. Do we have coverage set up?

Besides, if we would just start renaming internal functions to `s_*`, we wouldn't have to tweak the visibility settings in the first step. The `s_*` functions would just be internal and not belong to the public ABI by our own convention. In the next step we could make the symbols hidden.

> why is this too fragile?

The test of the internal function mp_balance_mul should not depend on the criterion at which this function is selected in mp_mul.

>  why would you want to test cases which can't be hit?

Robustness of the functions. I would not extend functions to accept values which cannot be handled fundamentally, but for example if karatsuba also works for small numbers and these should be tested too.

> that all sounds a bit over the top to me

No it is not. If you want to see something which is over the top look at the sqlite test suite. What I am proposing here is good practice and modest. It is also too much work hopefully. Furthermore there are people willing to improve stuff, like me getting this test suite refactoring started.

Reply:
> If only the external api should be tested, there should not even be a function test_balance_mul.

why not if this test is specifically crafted to test `mp_balance_mul` within `mp_mul`? 

> Do we have coverage set up?

you can do `make coverage` :)

> Besides, if we would just start renaming internal functions to `s_*`, we wouldn't have to tweak the visibility settings in the first step. The `s_*` functions would just be internal and not belong to the public ABI by our own convention. In the next step we could make the symbols hidden.

fine by me, keep the discussion points of #172 in mind

> The test of the internal function mp_balance_mul should not depend on the criterion at which this function is selected in mp_mul.

what sense does a specific test even make if its counterpart isn't available? as it is now the test's can't even be built w/o `mp_balance_mul`

> if karatsuba also works for small numbers and these should be tested too.

then set the cutoff in the testcase appropriately?